// @salesforce/plugin-login/lib/commands/login.js
"use strict";
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.LoginTarget = void 0;
const core_1 = require("@salesforce/core");
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
core_1.Messages.importMessagesDirectory(__dirname);
const messages = core_1.Messages.loadMessages('@salesforce/plugin-login', 'login');
var LoginTarget;
(function (LoginTarget) {
    LoginTarget["FUNCTIONS"] = "Salesforce Functions";
})(LoginTarget = exports.LoginTarget || (exports.LoginTarget = {}));
const LoginCommands = {
    [LoginTarget.FUNCTIONS]: 'login:functions',
};
class Login extends sf_plugins_core_1.SfCommand {
    async run() {
        return this.executeFunctionsLogin();
    }
    async executeFunctionsLogin() {
        await this.config.runCommand(LoginCommands[LoginTarget.FUNCTIONS]);
        return {};
    }
}
exports.default = Login;
Login.summary = messages.getMessage('summary');
Login.description = messages.getMessage('description');
Login.examples = messages.getMessages('examples');
Login.enableJsonFlag = false;
Login.flags = {};
Login.state = 'deprecated';
Login.deprecationOptions = {
    message: messages.getMessage('deprecationMessage'),
    version: '58.0',
};
