// @salesforce/plugin-deploy-retrieve/lib/utils/output.js
"use strict";
/*
 * Copyright (c) 2021, salesforce.com, inc.
 * All rights reserved.
 * Licensed under the BSD 3-Clause license.
 * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.check = exports.success = exports.error = exports.getFileResponseSuccessProps = exports.sortFileResponses = exports.asRelativePaths = exports.tableHeader = void 0;
const path = require("path");
const chalk_1 = require("chalk");
const sf_plugins_core_1 = require("@salesforce/sf-plugins-core");
function tableHeader(message) {
    return (0, chalk_1.blue)((0, chalk_1.bold)(message));
}
exports.tableHeader = tableHeader;
function asRelativePaths(fileResponses) {
    const relative = fileResponses.map((file) => file.filePath ? { ...file, filePath: path.relative(process.cwd(), file.filePath) } : file);
    return relative;
}
exports.asRelativePaths = asRelativePaths;
/**
 * Sorts file responds by type, then by filePath, then by fullName
 */
function sortFileResponses(fileResponses) {
    return fileResponses.sort((i, j) => {
        if (i.type === j.type && i.filePath && j.filePath) {
            if (i.filePath === j.filePath) {
                return i.fullName > j.fullName ? 1 : -1;
            }
            return i?.filePath > j?.filePath ? 1 : -1;
        }
        return i.type > j.type ? 1 : -1;
    });
}
exports.sortFileResponses = sortFileResponses;
const getFileResponseSuccessProps = (successes) => successes.map((s) => ({ filePath: s.filePath, fullName: s.fullName, type: s.type, state: s.state }));
exports.getFileResponseSuccessProps = getFileResponseSuccessProps;
function error(message) {
    return sf_plugins_core_1.StandardColors.error((0, chalk_1.bold)(message));
}
exports.error = error;
function success(message) {
    return sf_plugins_core_1.StandardColors.success((0, chalk_1.bold)(message));
}
exports.success = success;
exports.check = sf_plugins_core_1.StandardColors.success('✓');
